src/algebra/big_operators/finprod.lean

```diff
@@ -0,0 +1,623 @@
+  `f x ≠ 1`, if this is a finite sum. One otherwise.
+
+## Notation
+
+* `∑ᶠ i, f i` and `∑ᶠ i : α, f i` for `finsum f`
+
+* `∑ᶠ i in s, f i` for `finsum_in s f`
+
+* `∏ᶠ i, f i` and `∑ᶠ i : α, f i` for `finprod f`
+
+* `∏ᶠ i in s, f i` for `finprod_in s f`
+
+## Implementation notes
+
+`finsum` and `finprod` is "yet another way of doing finite sums and products in Lean". However
+experiments in the wild (e.g. with matroids) indicate that it is a helpful approach in settings
+where the user is not interested in computability and wants to do reasoning without running into
+typeclass diamonds caused by the constructive finiteness used in definitions such as `finset` and
+`fintype`. By sticking solely to `set.finite` we avoid these problems. We are aware that there are
+other solutions but for beginner mathematicians this approach is easier in practice.
+
+Another application is the construction of a partition of unity from a collection of “bump”
+function. In this case the finite set depends on the point and it's convenient to have a definition
+that does not mention the set explicitly.
+
+## Todo
+
+We did not add `is_finite (X : Type) : Prop`, because it is simply `nonempty (fintype X)`.
+There is work on `fincard` in the pipeline, which returns the cardinality of `X` if it
+is finite, and 0 otherwise.
+
+## Tags
+
+finsum, finsum_in, finprod, finprod_in, finite sum, finite product
+-/
+
+open_locale classical
+open function set
+
+/-!
+### Definition and relation to `finset.sum` and `finset.prod`
+-/
+
+section sort
+variables {α β ι : Sort*} {M N : Type*} [comm_monoid M] [comm_monoid N]
+
+open_locale big_operators
+
+/-- Sum of `f x` as `x` ranges over the elements of the support of `f`, if it's finite. Zero
+otherwise. -/
+@[irreducible] noncomputable def finsum {M} [add_comm_monoid M] (f : α → M) : M :=
+if h : finite (support (f ∘ plift.down)) then ∑ i in h.to_finset, f i.down else 0
+
+/-- Product of `f x` as `x` ranges over the elements of the multiplicative support of `f`, if it's
+finite. One otherwise. -/
+@[irreducible, to_additive]
+noncomputable def finprod (f : α → M) : M :=
+if h : finite (mul_support (f ∘ plift.down)) then ∏ i in h.to_finset, f i.down else 1
+
+localized "notation `∑ᶠ` binders `, ` r:(scoped:67 f, finsum f) := r" in big_operators
+
+localized "notation `∏ᶠ` binders `, ` r:(scoped:67 f, finprod f) := r" in big_operators
+
+@[to_additive] lemma finprod_eq_prod_plift_of_mul_support_to_finset_subset
+  {f : α → M} (hf : finite (mul_support (f ∘ plift.down))) {s : finset (plift α)}
+  (hs : hf.to_finset ⊆ s) :
+  ∏ᶠ i, f i = ∏ i in s, f i.down :=
+begin
+  rw [finprod, dif_pos hf],
+  refine finset.prod_subset hs (λ x hx hxf, _),
+  rwa [hf.mem_to_finset, nmem_mul_support] at hxf
+end
+
+@[to_additive] lemma finprod_eq_prod_plift_of_mul_support_subset
+  {f : α → M} {s : finset (plift α)} (hs : mul_support (f ∘ plift.down) ⊆ s) :
+  ∏ᶠ i, f i = ∏ i in s, f i.down :=
```

Comment:
How much would things clean up if `finset T` were allowed for `T : Sort*`?

Reply:
This would make the definitions more readable but I'm not sure that this is a simple task: there is a technical reason to have `pprod` separate from `prod` and `psigma` separate from `sigma` (I don't remember what is it). Anyway, I don't want this PR to depend on some large refactor of Lean core.